### PR TITLE
sliding sync: add receipts extension

### PR DESCRIPTION
**Pull request by kegsay**
## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * sliding sync: add receipts extension ([\#2912](https://github.com/matrix-org/matrix-js-sdk/pull/2912)).

### CODE REVIEW

File: spec/integ/sliding-sync-sdk.spec.ts [hunk 1]
```diff
@@ -928,4 +928,95 @@ describe("SlidingSyncSdk", () => {
...
+                                "m.read": {
+                                    [alice]: {
+                                        ts: 1234567,
+                                    },
+                                },
+                            },
+                        },
+                    },
+                },
+            });
```
Comment (alunturner) [inline spec/integ/sliding-sync-sdk.spec.ts:991]: The object that's used here is reused on line 1003. It looks to me like tests in this area might quite often have to generate this object, so I was wondering if you think it warrants making a tiny util function that takes in the `roomId,  eventId, username` and returns the required object?

Comment (kegsay) [reply spec/integ/sliding-sync-sdk.spec.ts:991]: I don't anticipate using this object a bunch, but given sonarcloud is whining I probably will be forced to make a util function for it.

File: src/sliding-sync-sdk.ts [hunk 2]
```diff
@@ -253,22 +253,48 @@ class ExtensionTyping implements Extension {
...
+        if (isInitial) {
+            return {
+                enabled: true,
+            };
+        }
+        return undefined;
+    }
+
+    public onResponse(data: {rooms: Record<string, IMinimalEvent>}): void {
+        if (!data || !data.rooms) {
```
Comment (richvdh) [inline src/sliding-sync-sdk.ts:290]: given empty objects are truthy... if the type annotation is correct then this condition will always fail?

File: src/sliding-sync-sdk.ts [hunk 3]
```diff
@@ -253,22 +253,48 @@ class ExtensionTyping implements Extension {
...
+        return ExtensionState.PostProcess;
+    }
+
+    public onRequest(isInitial: boolean): object | undefined {
+        if (isInitial) {
+            return {
+                enabled: true,
+            };
+        }
+        return undefined;
```
Comment (richvdh) [inline src/sliding-sync-sdk.ts:286]: I wouldn't have minded a comment to explain why this logic is correct, but that could be due to unfamiliarity with sliding sync and the modules therein

**Review by richvdh — APPROVED** [review]
seems generally plausible

**Pull request merged**